pkg/rate/limiter.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rate
```

Comment:
Is there a reason to not use https://godoc.org/golang.org/x/time/rate ?

Reply:
Yes, there is a reason. We were initially using `x/time/rate`, but it turns out that the token bucket algorithm used by `x/time/rate` is not suitable for our purpose.

An observation we made in reproducing #8360 is that even when there are thousands of identities marked for deletion, cilium-operator already does not seem to produce more than a few dozen deletion events per second (since it has to do a kvstore operation for each identity). 

The problem is that it sends a few dozen deletion events per second **over the period of a few minutes**. This in turn causes cilium-agent to constantly regenerate the endpoints (as it is receiving new updates every second), causing it to become unresponsive while cilium-operator is deleting the identities.

To formulate simply: The main problem are not mainly bursts, the problem is the constant stream of updates.

So a token bucket algorithm as implemented by `x/time/rate` will not fix the issue. It is designed to flatten bursts in a signal to a fixed output rate:

![image](https://user-images.githubusercontent.com/50564/86932967-51a12280-c13a-11ea-8382-63db93315e98.png)

Our main concern however are not bursts. In fact, Cilium is very able to deal with small bursts of identity changes, as it does coalesce identity updates happening within a predefined interval (`--policy-trigger-interval`, default 1s). 

Therefore, André's custom rate limiter kind of does the opposite of `x/time/rate`. It takes a somewhat fixed-rate stream of updates and turns it into a stream of controlled small bursts (250 by default) every 1 minutes. 

This gives cilium-agent time to recover from endpoint regeneration churn within the one minute idle time.

@aanm: We probably should add a comment in the source code somewhere about so nobody accidentally replaces the rate limiter with `x/time/rate`.

Reply:
+1 on the comment. Heck, even this explanation by @gandro would be great to have as the package documentation as well. Thanks for the great explanation @gandro , very easy to understand.

Reply:
@gandro I've had it here, if't not enough what other places should I put it? https://github.com/cilium/cilium/blob/15adbd708765c2c01029ea2d622b1f2c79241a63/pkg/rate/doc.go#L15-L18

Reply:
The comment explains the "what", but I think we are missing the "why" we are using that algorithm for identity GC.

It's not obvious from the identitiy GC code that the traffic pattern produced by rate limiting algorithm matters. We currently only explain is how it is different from `golang.org/x/time/rate`, but not why it needs to be different.

To be clear, the "what it does" package level doc in `rate` that you quoted is fine with me. But I think we should add a comment here: https://github.com/cilium/cilium/blob/70d561786228892a2988ccccbe5ed9b827986acf/operator/main.go#L77
There we should have a comment about "why" we need a algorithm that produces small-bursts with pauses in between.

Reply:
We can add a comment in a follow up. I think documenting the "why" is very valuable to our future selves and newcomers who are taking a look at this code.